### PR TITLE
[TreeView] Improve the selection API

### PR DESCRIPTION
**Pull request by flaviendelangle**
Follow up on #11476

- [x] Package changes
- [x] Docs changes (the demo for `onNodeExpansionToggle` will come after the doc split of #11059
- [x] Migration guide
- [x] Codemod

I have one issue with the PR as it is.
The selection model contains a single nodeId (its type is `string`) when `multiSelect={false}`.
So naming it `selectedNodes` in plural can be misleading.
On the other hand, keeping `selected` feels like a bad solution...

If you have a better idea :grimacing:

### CODE REVIEW

File: packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.types.ts [hunk 1]
```diff
@@ -26,32 +26,43 @@ export interface UseTreeViewSelectionParameters<Multiple extends boolean | undef
...
-  selected?: TreeViewSelectionValue<Multiple>;
+  selectedNodes?: TreeViewSelectionValue<Multiple>;
   /**
    * If true `ctrl` and `shift` will trigger multiselect.
    * @default false
    */
   multiSelect?: Multiple;
   /**
-   * Callback fired when tree items are selected/unselected.
+   * Callback fired when tree items are selected/deselected.
```
Comment (flaviendelangle) [inline packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.types.ts:41]: See https://english.stackexchange.com/questions/18465/unselect-or-deselect#:~:text=The%20NOAD%20defines%20deselect%20as,selected%20and%20isn't%20anymore.

File: docs/data/tree-view/rich-tree-view/selection/selection.md [hunk 2]
```diff
@@ -32,3 +32,9 @@ You can use the `onNodeSelect` prop to listen to changes in the selected items a
 
 Learn more about the _Controlled and uncontrolled_ pattern in the [React documentation](https://react.dev/learn/sharing-state-between-components#controlled-and-uncontrolled-components).
 :::
+
+## Track node selection change
+
+Use the `onNodeSelectionToggle` if you want to react to a node selection toggling:
```
Comment (michelengelen) [inline docs/data/tree-view/rich-tree-view/selection/selection.md:38]: not sure if this is a valid change request, but definitely a nit:

```suggestion
Use the `onNodeSelectionToggle` if you want to react to a node selection toggle:
```

🤷🏼

File: docs/data/tree-view/simple-tree-view/selection/selection.md [hunk 3]
```diff
@@ -32,3 +32,9 @@ You can use the `onNodeSelect` prop to listen to changes in the selected items a
 
 Learn more about the _Controlled and uncontrolled_ pattern in the [React documentation](https://react.dev/learn/sharing-state-between-components#controlled-and-uncontrolled-components).
 :::
+
+## Track node selection change
+
+Use the `onNodeSelectionToggle` if you want to react to a node selection toggling:
```
Comment (michelengelen) [inline docs/data/tree-view/simple-tree-view/selection/selection.md:38]: same here as in `docs/data/tree-view/rich-tree-view/selection/selection.md`

File: packages/x-codemod/README.md [hunk 4]
```diff
@@ -238,6 +241,40 @@ Renames the `TreeView` component to `SimpleTreeView`
...
+  <TreeView
+-   onNodeSelect={handleSelectionChange}
++   onSelectedNodesChange={handleSelectionChange}
+
+-   selected={selectedNodes}
++   selectedNodes={selectedNodes}
+
+-   defaultSelected={defaultSelectedNodes}
++   defaultSelectedNodes={defaultSelectedNodes}
+  />
```
Comment (michelengelen) [inline packages/x-codemod/README.md:258]:
```suggestion
  <TreeView
-   onNodeToggle={handleExpandedNodesChange}
+   onExpandedNodesChange={handleExpandedNodesChange}

-   expanded={expandedNodes}
+   expandedNodes={expandedNodes}

-   defaultExpanded={defaultExpandedNodes}
+   defaultExpandedNodes={defaultExpandedNodes}
  />
```

**Review by michelengelen — COMMENTED** [review]
looking good ... just a nit and something that got copied and pasted :D

File: packages/x-codemod/README.md (hunk 4 above)
Comment (flaviendelangle) [reply packages/x-codemod/README.md]: Oops, thanks for catching that one!

**Review by michelengelen — APPROVED** [review]

**Review by LukasTy — APPROVED** [review]
LGTM. 👍 
Nice improvement. 💯

**Pull request merged**